src/simulations/set_generator.py

```diff
@@ -626,3 +626,96 @@ def __iter__(self):
+  Heterogeneous means that the reached ID's have different frequency
+  distribution.
+
+  As a reached ID may appear one or more times in the returned set,
+  to differentiate with a normal set whose elements can only appear once, we
+  use the term multiset.
+
+  The frequency distribution is defined as a shifted-Poisson distribution.
+  I.e., freq ~ Poission(freq_rate) + 1.
+  """
+
+  @classmethod
+  def get_generator_factory_with_num_and_size(cls, universe_size, num_sets,
+                                              set_size, gamma_params_list,
+                                              freq_cap):
+
+    def f(random_state):
+      return cls(universe_size, list(_SetSizeGenerator(num_sets, set_size)),
+                 gamma_params_list, random_state, freq_cap)
+
+    return f
+
+  @classmethod
+  def get_generator_factory_with_set_size_list(cls, universe_size,
+                                               set_size_list, gamma_params_list,
+                                               freq_cap):
+
+    def f(random_state):
+      return cls(universe_size, set_size_list, gamma_params_list, random_state,
+                 freq_cap)
+
+    return f
+
+  def __init__(self, universe_size, set_sizes, gamma_params_list, random_state,
+               freq_cap=None):
+    """Create a heterogeneous multiset generator.
+
+    Args:
+      universe_size: An integer value that specifies the size of the whole
+        universe from which the ids will be sampled.
+      set_sizes: A list containing the size of each set, aka, the number of
+        reached IDs.
+      gamma_params_list: A list of the same size as set_sizes, specifying the
+        non-negative gamma_rate parameter of the gamma distribution to sample
+        lambda parameter for that set's shifted-Possion distribution.
+      random_state: a numpy random state.
+      freq_cap: An optional positive integer which represents the maximum number
+        of times an ID can be seen in the returned set. If not set, will not
+        apply this capping.
+
+    Raises: AssertionError when (1) set_size_list and gamma_params_list do not
+      have equal length, (2) elements of gamma_params_list are not all
+      non-negative, or (3) freq_cap is not None or positive.
+    """
+    assert len(set_sizes) == len(gamma_params_list), (
+        'set_size_list and gamma_params_list do not have equal length.')
+    assert all([gamma_params[0] > 0 for gamma_params in gamma_params_list]), (
+        'Gamma shape parameters must be positive.')
+    assert all([gamma_params[1] > 0 for gamma_params in gamma_params_list]), (
+        'Gamma rate parameters must be positive.')
+    assert freq_cap is None or freq_cap > 0, (
+        'freq_cap should be None or positive.')
+    self.universe_size = universe_size
+    self.set_sizes = set_sizes
+    self.gamma_params_list = gamma_params_list
+    self.freq_cap = freq_cap
+    self.random_state = random_state
+    self.choice = choice_fast
```

Comment:
Previously "self.choice" existed so we could fake the choice algorithm in test code. Since you are not faking it I think it's reasonable to just call "choice_fast" in __iter__.

Reply:
Done.